includes/handlers/class-webmention-handler-meta.php

```diff
@@ -0,0 +1,224 @@
+ * Class for webmention parsing using META tags.
+ */
+class Webmention_Handler_Meta extends Webmention_Handler_Base {
+
+	/**
+	 * Handler Slug to Uniquely Identify it.
+	 *
+	 * @var string
+	 */
+	protected $slug = 'meta';
+
+	/**
+	 * Takes a request object and parses it.
+	 *
+	 * @param Webmention_Request $request Request Object.
+	 * @param Webmention_Item $item A Parsed Item. If null, a new one will be created.
+	 * @return WP_Error|true Return error or true if successful.
+	 */
+	public function parse( $request, $item = null ) {
+		if ( $item instanceof Webmention_Item ) {
+			$this->webmention_item = $item;
+		} else {
+			$this->webmention_item = new Webmention_Item();
+		}
+		$dom  = clone $request->get_domdocument();
+		$meta = $this->parse_meta( $dom );
+
+		// Set raw data.
+		$this->webmention_item->set__raw( $meta );
+
+		// OGP has no concept of anything but mention so it is always a mention.
+		$this->webmention_item->set__response_type = 'mention';
+
+		$this->ogp_to_jf2( $meta );
+		$this->dublin_core_to_jf2( $meta );
+
+		// If Site Name is not set use domain name less www
+		if ( ! $this->webmention_item->has_site_name() && $this->webmention_item->has_url() ) {
+			$this->webmention_item->set__site_name( preg_replace( '/^www\./', '', wp_parse_url( $this->webmention_item->get_url(), PHP_URL_HOST ) ) );
+		}
+
+		$this->webmention_item->set_name( $meta['title'] );
+
+		if ( ! $this->webmention_item->has_published() ) {
+			if ( isset( $meta['citation_date'] ) ) {
+				$this->webmention_item->set_published( new DateTimeImmutable( $meta['citation_date'] ) );
+			} elseif ( isset( $meta['datePublished'] ) ) {
+				$this->webmention_item->set_published( new DateTimeImmutable( $meta['datePublished'] ) );
+			}
+		}
+
+		if ( ! $this->webmention_item->has_author() ) {
+			$this->webmention->item->set_author( array( 'name' => $meta['author'] ) );
+		}
+	}
+
+	/**
+	 * Takes a DOMDocument and returns an array of parsed meta properties.
+	 *
+	 * @param DOMDocument $dom DOMDocument.
+	 * @return array Associative array.
+	 */
+	private function parse_meta( $dom ) {
+		$xpath = new DOMXPath( $dom );
+		$meta  = array();
+		// Look for OGP properties
+		foreach ( $xpath->query( '//meta[(@name or @property or @itemprop) and @content]' ) as $tag ) {
+			$meta_name = $tag->getAttribute( 'property' );
+			if ( ! $meta_name ) {
+				$meta_name = $tag->getAttribute( 'name' );
+			}
+			if ( ! $meta_name ) {
+				$meta_name = $tag->getAttribute( 'itemprop' );
+			}
+			$meta_value = $tag->getAttribute( 'content' );
+			// Sanity check. $key is usually things like 'title', 'description', 'keywords', etc.
+			if ( strlen( $meta_name ) > 200 ) {
+				continue;
+			}
+			$meta = $this->add_property( $meta, $meta_name, $meta_value );
+		}
+
+		$meta['title'] = trim( $xpath->query( '//title' )->item( 0 )->textContent );
+		$meta          = $this->parse_ogp( $meta );
+		if ( isset( $meta['og'] ) ) {
+			$meta['og'] = $this->parse_ogp( $meta['og'] );
```

Comment:
Does this really work? it seems to result in `$meta['og']['og']` or am I wrong? If `parse_ogp` also interprets `DC` it will also result in `$meta['og']['dc']`.

Reply:
If you look at the code, it is deliberate as it tries to separate out the hyphenated properties into a property hierarchy....so it is parsing it to an array structure.